dandi/validate_types.py

```diff
@@ -1,21 +1,83 @@
 from __future__ import annotations
 
-from dataclasses import dataclass
-from enum import Enum
+from enum import Enum, IntEnum, auto, unique
 from pathlib import Path
+from typing import Any
 
+from pydantic import BaseModel, Field
 
-@dataclass
-class ValidationOrigin:
-    name: str
-    version: str
-    bids_version: str | None = None
+from dandi.utils import StrEnum
 
 
-class Severity(Enum):
-    HINT = 1
-    WARNING = 2
-    ERROR = 3
+@unique
+class Standard(StrEnum):
+    """Standards to validate against"""
+
+    BIDS = auto()
+    DANDI_LAYOUT = "DANDI-LAYOUT"
+    DANDI_SCHEMA = "DANDI-SCHEMA"
+    HED = auto()
+    NWB = auto()
+    OME_ZARR = "OME-ZARR"
+    ZARR = auto()
+
+    # File formats (For denoting validation failures in file format level)
+    JSON = auto()
+    TSV = auto()
+    YAML = auto()
+
+
+@unique
+class Validator(StrEnum):
+    """Validators that are used to do validation"""
+
+    bids_validator_deno = "bids-validator-deno"
+    bidsschematools = auto()
+    dandi = auto()
+    dandi_zarr = "dandi.zarr"
+    dandischema = auto()
+    hed_python_validator = "hed-python-validator"
+    nwbinspector = auto()
+    pynwb = auto()
+    zarr = auto()
+
+
+class Origin(BaseModel):
+    validator: Validator
+    """The validator conducting the validation"""
+
+    validator_version: str
+    """The version of the validator"""
+
+    standard: Standard | None = None
+    """Standard being validated against"""
+
+    standard_version: str | None = None
+    """Version of the standard"""
```

Comment:
may be here it could be `source` which could be "standard" or "internal" ATM, but then may be smth else too

Reply:
Since this is a field in the `Origin` object. I think naming the field `source` is a bit confusing.

Since we already use an `Origin` object to represent the "source" of a validation result, and, at this point, we want to indicate rather the "source" is an "external" validator invoked by the `dandi-cli` or some "internal" code of the `dandi-cli` itself, I think we can name the field in the `Origin` object `type` and allow the value of the field to be `internal`, `external`, or `undetermined`. The `undetermined` value is needed because there are cases where source of the error is hard/impossible to determine.

Reply:
I have decided on types of the origin to be 

```py
class Type(StrEnum):
    """Types of validation result origins"""

    INTERNAL = auto()
    """Validation result is originated from the code of this project"""

    VALIDATOR = auto()
    """Validation result is originated from a validator"""

    UNDETERMINED = auto()
    """
    Validation result is originated from a source that is not determined to
    be another value in this enum class
    
    This is usually used when the origin of the validation result is unclear
    """
```

I used `VALIDATOR` instead of your suggestion of `STANDARD` because not all validator implements a standard (nwbinspector, for example).